File: cses/06_tree_algorithms/2079_finding-a-centroid/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    idx = 0
    n = int(data[idx]); idx += 1

    # Trường hợp đặc biệt: cây một đỉnh, centroid là đỉnh 1.
    if n == 1:
        sys.stdout.buffer.write(b"1\n")
        return

    # Đọc n-1 cạnh vô hướng và lưu cây bằng adjacency list.
    adj = [[] for _ in range(n + 1)]
    for _ in range(n - 1):
        a = int(data[idx]); b = int(data[idx + 1]); idx += 2
        adj[a].append(b)
        adj[b].append(a)

    # Iterative DFS từ gốc 1 (không đệ quy) để tính parent và thứ tự preorder.
    parent = [0] * (n + 1)
    order = []
    visited = [False] * (n + 1)
    stack = [1]
    visited[1] = True
    while stack:
        u = stack.pop()
        order.append(u)
        for v in adj[u]:
            if not visited[v]:
                visited[v] = True
                parent[v] = u
                stack.append(v)

    # Tính size[u] = số đỉnh trong subtree gốc u bằng cách duyệt ngược preorder.
    size = [1] * (n + 1)
    for u in reversed(order):
        p = parent[u]
        if p != 0:
            size[p] += size[u]

    # Centroid là đỉnh đầu tiên mà mọi thành phần khi xóa nó đều <= floor(n/2):
    # phần "phía trên" (n - size[u]) và từng subtree con (size[v]).
    half = n // 2
    centroid = 1
    for u in range(1, n + 1):
        if n - size[u] > half:
            continue
        ok = True
        for v in adj[u]:
            if parent[v] == u and size[v] > half:
                ok = False
                break
        if ok:
            centroid = u
            break

    sys.stdout.buffer.write(str(centroid).encode() + b"\n")


main()

File: cses/06_tree_algorithms/2079_finding-a-centroid/solution.cpp
#include <algorithm>
#include <iostream>
#include <vector>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n;
    std::cin >> n;

    // Đọc n-1 cạnh vô hướng và lưu cây bằng adjacency list.
    std::vector<std::vector<int>> graph(n + 1);
    for (int edge = 0; edge < n - 1; ++edge) {
        int first, second;
        std::cin >> first >> second;
        graph[first].push_back(second);
        graph[second].push_back(first);
    }

    // Iterative DFS từ gốc 1 (dùng stack tường minh, không đệ quy) để tính
    // parent của mỗi đỉnh và thứ tự duyệt preorder.
    std::vector<int> parent(n + 1, 0);
    std::vector<int> order;
    order.reserve(n);
    std::vector<int> stack{1};
    parent[1] = -1;

    while (!stack.empty()) {
        const int node = stack.back();
        stack.pop_back();
        order.push_back(node);
        for (const int next : graph[node]) {
            if (next == parent[node]) {
                continue;
            }
            parent[next] = node;
            stack.push_back(next);
        }
    }

    // Tính size[u] = số đỉnh trong subtree gốc u bằng cách duyệt ngược
    // preorder và cộng dồn kích thước con lên cha.
    std::vector<int> subtree_size(n + 1, 1);
    for (auto iterator = order.rbegin(); iterator != order.rend(); ++iterator) {
        const int node = *iterator;
        if (parent[node] > 0) {
            subtree_size[parent[node]] += subtree_size[node];
        }
    }

    // Với mỗi đỉnh, tìm thành phần liên thông lớn nhất khi xóa nó: gồm phần
    // "phía trên" (n - size[node]) và mỗi subtree con (size[next]).
    // Centroid là đỉnh đầu tiên có thành phần lớn nhất <= floor(n/2).
    const int half = n / 2;
    for (int node = 1; node <= n; ++node) {
        int largest_component = n - subtree_size[node];
        for (const int next : graph[node]) {
            if (parent[next] == node) {
                largest_component = std::max(
                    largest_component, subtree_size[next]
                );
            }
        }
        if (largest_component <= half) {
            std::cout << node << '\n';
            break;
        }
    }
    return 0;
}
